From b2c7c8d50d0443b759cdd8d84e307fea9b26f991 Mon Sep 17 00:00:00 2001
From: Philippe Normand <philn@igalia.com>
Date: Wed, 10 Aug 2016 12:28:52 +0200
Subject: [PATCH] optional glvideoflip

---
 local/owr_video_renderer.c | 23 ++++++++++++++++-------
 1 file changed, 16 insertions(+), 7 deletions(-)

diff --git a/local/owr_video_renderer.c b/local/owr_video_renderer.c
index d8898de..1a379e4 100644
--- a/local/owr_video_renderer.c
+++ b/local/owr_video_renderer.c
@@ -311,10 +311,13 @@ static GstElement *owr_video_renderer_get_element(OwrMediaRenderer *renderer, gu
     renderer_disabled(renderer, NULL, balance);
 
     flip = gst_element_factory_make("glvideoflip", "video-renderer-flip");
-    g_assert(flip);
-    g_signal_connect_object(renderer, "notify::rotation", G_CALLBACK(update_flip_method), flip, 0);
-    g_signal_connect_object(renderer, "notify::mirror", G_CALLBACK(update_flip_method), flip, 0);
-    update_flip_method(renderer, NULL, flip);
+    if (!flip) {
+        g_warning("The glvideoflip element isn't available");
+    } else {
+        g_signal_connect_object(renderer, "notify::rotation", G_CALLBACK(update_flip_method), flip, 0);
+        g_signal_connect_object(renderer, "notify::mirror", G_CALLBACK(update_flip_method), flip, 0);
+        update_flip_method(renderer, NULL, flip);
+    }
 
     sink = OWR_MEDIA_RENDERER_GET_CLASS(renderer)->get_sink(renderer);
     g_assert(sink);
@@ -330,12 +333,18 @@ static GstElement *owr_video_renderer_get_element(OwrMediaRenderer *renderer, gu
             g_object_unref(sink_element);
     }
 
-    gst_bin_add_many(GST_BIN(renderer_bin), upload, convert, balance, flip, sink, NULL);
+    gst_bin_add_many(GST_BIN(renderer_bin), upload, convert, balance, sink, NULL);
 
     LINK_ELEMENTS(upload, convert);
     LINK_ELEMENTS(convert, balance);
-    LINK_ELEMENTS(balance, flip);
-    LINK_ELEMENTS(flip, sink);
+
+    if (flip) {
+        gst_bin_add(GST_BIN(renderer_bin), flip);
+        LINK_ELEMENTS(balance, flip);
+        LINK_ELEMENTS(flip, sink);
+    } else {
+        LINK_ELEMENTS(balance, sink);
+    }
 
     sinkpad = gst_element_get_static_pad(upload, "sink");
     g_assert(sinkpad);
-- 
2.8.1

